### PR TITLE
chore: generate python bindings with better types

### PR DESCRIPTION
**Pull request by rb-determined-ai**
The generated bindings are now generous about what they accept, but
strict in what they return (from a typing perspective).

In cases where an input type annotation (like Sequence) is converted to
an output type annotation (like List), the originally-provided value
will be used if it matches the output type, otherwise a new object of
the stricter type will be created.

## Commentary (optional)

@mpkouznetsov was curious why there were some extra calls to `list()` when using the generated bindings.  There were cases where a Sequence was being returned, which was not very helpful to the caller of the bindings.

### CODE REVIEW

File: bindings/generate_bindings.py [hunk 1]
```diff
@@ -12,7 +13,16 @@
 
 
 class TypeAnno:
-    def annotation(self, prequoted=False) -> Code:
+    def input_annotation(self, prequoted=False) -> Code:
+        """input_annotation is for parameters; things accepted from a user."""
+        raise NotImplementedError(type(self))
+
+    def output_annotation(self, prequoted=False) -> Code:
+        """input_annotation is for return types and attributes; things provided to a user."""
```
Comment (mpkouznetsov) [inline bindings/generate_bindings.py:21]: you mean output annotation?

File: bindings/generate_bindings.py [hunk 2]
```diff
@@ -301,7 +351,9 @@ def gen_def(self) -> Code:
         for name in required + optional:
             out += ["    " + self.params[name].gen_function_param()]
         out += ["    ):"]
-        out += [f"        self.{k} = {k}" for k in self.params]
+        for name in required + optional:
+            p = self.params[name]
+            out += [f"        self.{name} = {p.type.input_to_output(name, not p.required)}"]
```
Comment (rb-determined-ai) [inline bindings/generate_bindings.py:356]: Note that before, the order of class `__init__` parameters did not match the order in which parameters were saved to the `self` object, which was annoying when inspecting the generated code.

I changed it, but it also caused a lot more churn in the generated bindings, sorry.

Comment (rb-determined-ai) [reply bindings/generate_bindings.py:356]: this is now the second two commits, to make reviewing easier.

**Review by hamidzr — APPROVED** [review]
lgtm

we could add this as demo for the change
```
diff --git a/harness/determined/cli/experiment.py b/harness/determined/cli/experiment.py
index b654d61be..113f1d32d 100644
--- a/harness/determined/cli/experiment.py
+++ b/harness/determined/cli/experiment.py
@@ -688,7 +688,7 @@ def add_label(args: Namespace) -> None:
     if exp_patch.labels is None:
         exp_patch.labels = []
     if args.label not in exp_patch.labels:
-        exp_patch.labels = list(exp_patch.labels) + [args.label]
+        exp_patch.labels = exp_patch.labels + [args.label]
         bindings.patch_PatchExperiment(session, body=exp_patch, experiment_id=args.experiment_id)
     print("Added label '{}' to experiment {}".format(args.label, args.experiment_id))
 
```

**Comment by rb-determined-ai** [comment]
test failure is real, it'll take some investigation.

**Comment by rb-determined-ai** [comment]
closing due to staleness

**Pull request closed without merging**